### PR TITLE
Add acts as taggable on

### PR DESCRIPTION
**Pull request by lzhang0616**

### CODE REVIEW

**Comment by lzhang0616** [comment]
I haven't fix them all in #5..... Once I added 'git config --global core.autocrlf true'  for my git global setting after I read about the [Ruby-coding-style](https://github.com/bbatsov/ruby-style-guide). So my several files use CRLF for new line....

**Comment by lzhang0616** [comment]
We decided to put the update link in the show page, when I paired with Jim this afternoon....

**Comment by rmacklin** [comment]
> We decided to put the update link in the show page, when I paired with Jim this afternoon

I see. That is probably due to a miscommunication between us and the pairing mentors. I don't think we mentioned that we had intentionally not postponed adding an `update` action until a further user story. We will try to do better about letting the pairing mentors know about the planned stories.

**Comment by lzhang0616** [comment]
So where should I put this update link? @rmacklin

**Comment by rmacklin** [comment]
You could add a field for tags on the `new` page.

**Comment by lzhang0616** [comment]
Oh, I see..... I will make this change now. And I almost figured out how to do the next story now. After I move this field into new page, I will continue to work on the next story.

**Comment by lzhang0616** [comment]
I will show the hashtag as a link to all the images has this hashtag in both index page and show page, is that ok? @rmacklin

**Comment by lzhang0616** [comment]
Can we create a new image with no tag? @rmacklin

**Comment by lzhang0616** [comment]
Now the image by tag links works for my app

**Comment by lzhang0616** [comment]
Since now we have communication between two controllers (ImagesController and TagsController), we do need to add some integration tests to make sure all the redirections work. However, we may have to add some CSS styles before we can write these tests, since we do need some handler for integration tests.

**Comment by rmacklin** [comment]
> However, we may have to add some CSS styles before we can write these tests, since we do need some handler for integration tests.

That's not exactly true. You can add IDs and classes to elements without styling them, and use those attributes to select the elements in your capybara tests.

**Comment by lzhang0616** [comment]
Yeah, I see. We can also use Xpath :)

**Comment by bboe** [comment]
Now that we have issues, can you please add "Resolves issue #?" in the pull request description?

File: db/schema.rb [hunk 1]
```diff
@@ -19,4 +19,24 @@
     t.datetime "updated_at", null: false
   end
 
+  create_table "taggings", force: :cascade do |t|
+    t.integer  "tag_id"
```
Comment (bboe) [inline db/schema.rb]: Do all of these, and those in table `tags`, need to be `null: true`?

I know this is what acts-as-taggable sets up, but I'm guessing some of these should never be null.

File: app/assets/stylesheets/tags.css [hunk 2]
```diff
@@ -0,0 +1,2 @@
+.tag_link{}
```
Comment (bboe) [inline app/assets/stylesheets/tags.css]: What do we gain by this addition?

File: app/controllers/images_controller.rb [hunk 3]
```diff
@@ -1,38 +1,42 @@
...
+  def new
+    @image = Image.new
+  end
+
+  def create
+    url = params[:image][:url]
+    tag_list = params[:image][:tag_list]
+
+    @image = Image.new(url: url)
+    @image.tag_list &&= tag_list
```
Comment (bboe) [inline app/controllers/images_controller.rb]: Can we pass the tag list directly into `Image.new`?

File: app/assets/stylesheets/tags.css (hunk 2 above)
Comment (lzhang0616) [inline app/assets/stylesheets/tags.css]: This is used to select the images with the specific tag in the tag's show page in TagsControllerTest

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -1,80 +1,98 @@
...
+  test 'create an image with tags successfully' do
+    assert_difference('Image.count', 1) do
+      post :create, image: { url: VALID_IMAGE_URL, tag_list: 'awsome, good' }
+    end
+
+    assert_equal %w(awsome good), Image.last.tag_list
+    assert_redirected_to image_path(assigns(:image))
+  end
+
+  test 'create an image omitting white space tag' do
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: Is it possible to create a test with an invalid tag list? E.g., what if it's not a string, or it's an array.

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -1,2 +1,9 @@
 module ApplicationHelper
+  def render_images(images)
```
Comment (bboe) [inline app/helpers/application_helper.rb]: If we were to add 20 models each with similar helpers, what might happen to the application helper?

Is there some other place we could put these helpers?

File: test/controllers/images_controller_test.rb (hunk 4 above)
Comment (lzhang0616) [inline test/controllers/images_controller_test.rb]: Sure, I will add these tests.

File: app/controllers/images_controller.rb (hunk 3 above)
Comment (lzhang0616) [inline app/controllers/images_controller.rb]: The tag list is directly pass into Image.new now.

File: app/helpers/application_helper.rb (hunk 5 above)
Comment (lzhang0616) [inline app/helpers/application_helper.rb]: These two helpers are used both by ImagesController and TagsController. I have tried, if I put them into ImagesHelper or TagsController, then it does not work.

File: app/assets/stylesheets/tags.css (hunk 2 above)
Comment (bboe) [inline app/assets/stylesheets/tags.css]: Try removing this file and see if it still works :).

File: app/helpers/application_helper.rb (hunk 5 above)
Comment (bboe) [inline app/helpers/application_helper.rb]: What you described should work. See http://api.rubyonrails.org/classes/ActionController/Helpers.html

File: app/views/images/new.html.erb [hunk 6]
```diff
@@ -1,7 +1,9 @@
 <h1>Please add image URL below</h1>
 
 <%= form_for @image do |f| %>
-    <%= f.label :url %>:
-    <%= f.text_field :url %><br />
-    <%= f.submit %>
+  <%= f.label :url %>:
+  <%= f.text_field :url %><br />
+  <%= f.label :tags %>:
+  <%= f.text_field :tag_list %><br />
```
Comment (bboe) [inline app/views/images/new.html.erb]: In html5 the `<br>` tag doesn't require the trailing `/`.

File: app/helpers/application_helper.rb (hunk 5 above)
Comment (lzhang0616) [inline app/helpers/application_helper.rb]: Ok, I see. I will make this change.

File: app/assets/stylesheets/tags.css (hunk 2 above)
Comment (lzhang0616) [inline app/assets/stylesheets/tags.css]: Aha! It still works :) I learned that CSS classes are not required to be pre-defined as programming language variables. Thanks!

File: db/schema.rb (hunk 1 above)
Comment (lzhang0616) [inline db/schema.rb]: Yeah, I think tag_id and the tag's referring class's (in our case the image) id and type should never be null.

File: app/views/images/new.html.erb (hunk 6 above)
Comment (lzhang0616) [inline app/views/images/new.html.erb]: Ok, I will remove it.

File: test/controllers/tags_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  VALID_IMAGE_URL = 'https://upload.wikimedia.org/wikipedia/en/thumb/e/e9/Ruby_on_Rails.svg/791px-Ruby_on_Rails.svg.png'
+
+  setup do
+    2.times { create_image_with_tag('red') }
+    2.times { create_image_with_tag('green') }
+    2.times { create_image_with_tag('blue') }
+    2.times { create_image_with_tag('red, blue') }
+  end
+
+  test 'show all images with tag #green' do
```
Comment (bboe) [inline test/controllers/tags_controller_test.rb]: Great tests. What's the one test that's missing?

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -1,80 +1,98 @@
...
+  VALID_IMAGE_URL = 'https://upload.wikimedia.org/wikipedia/en/thumb/e/e9/Ruby_on_Rails.svg/791px-Ruby_on_Rails.svg.png'
+
+  test 'should get index' do
+    get :index
+    assert_response :success
+  end
+
+  test 'create a new image' do
+    assert_difference('Image.count', 1) do
+      post :create, image: { url: VALID_IMAGE_URL }
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: What's missing from this test?

File: test/controllers/images_controller_test.rb [hunk 9]
```diff
@@ -1,80 +1,98 @@
...
+    end
+
+    assert_redirected_to image_path(assigns(:image))
+  end
+
+  test 'fail to create a new image' do
+    assert_no_difference('Image.count') do
+      post :create, image: { url: '' }
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: What's missing from this test?

File: test/controllers/tags_controller_test.rb (hunk 7 above)
Comment (lzhang0616) [inline test/controllers/tags_controller_test.rb]: I think we need to add a test for a non-exist tag, and another for tag red.

File: app/controllers/tags_controller.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+class TagsController < ApplicationController
```
Comment (bboe) [inline app/controllers/tags_controller.rb]: Let's leave this out of this PR. We'll have a discussion about how to filter the image list by tag, but let's not do it here.

Comment (lzhang0616) [inline app/controllers/tags_controller.rb]: Sure

**Comment by shiboying** [comment]
I checked on review app and adding a tag/multi-tags works well :+1:

**Comment by lzhang0616** [comment]
Closes #13

**Pull request merged**